From f513d278d4d9e51d6786512d2e3cc0d3bd89b838 Mon Sep 17 00:00:00 2001
From: Andoni Morales Alastruey <ylatuya@gmail.com>
Date: Sun, 25 Jul 2010 17:04:12 +0200
Subject: [PATCH 2/2] dshowvideosink: Close our own window when changing the window id

If we created a window, it needs to be closed after setting a new
window id.
---
 sys/dshowvideosink/dshowvideosink.cpp |   14 ++++++++++----
 1 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/sys/dshowvideosink/dshowvideosink.cpp b/sys/dshowvideosink/dshowvideosink.cpp
index bfc1c25..9fe5bf5 100644
--- a/sys/dshowvideosink/dshowvideosink.cpp
+++ b/sys/dshowvideosink/dshowvideosink.cpp
@@ -122,10 +122,16 @@ gst_dshowvideosink_set_window_id (GstXOverlay * overlay, ULONG window_id)
   /* Update window if we're already playing. */
   if (sink->connected && sink->filter_media_event) {
     HRESULT hres;
-
-    /* Return control of application window */
-    SetWindowLongPtr (previous_window, GWL_WNDPROC, (LONG)sink->prevWndProc);
-    SetWindowPos (previous_window, 0, 0, 0, 0, 0, SWP_NOMOVE | SWP_NOSIZE | SWP_NOZORDER | SWP_FRAMECHANGED);
+ 		
+    if (sink->is_new_window) {
+      /* If we created a new window */
+      SendMessage (sink->window_id, WM_CLOSE, NULL, NULL);
+      sink->is_new_window = FALSE;
+    } else {
+      /* Return control of application window */
+      SetWindowLongPtr (previous_window, GWL_WNDPROC, (LONG)sink->prevWndProc);
+      SetWindowPos (previous_window, 0, 0, 0, 0, 0, SWP_NOMOVE | SWP_NOSIZE | SWP_NOZORDER | SWP_FRAMECHANGED);
+    }
 
     gst_dshowvideosink_set_window_for_renderer (sink);
 
-- 
1.7.0.2.msysgit.0

